### ISSUE TITLE
Unhandled capturing of generic types within nested Job Expressions

### ISSUE THREAD

**Issue by stefanmoosbrugger** [insieme/insieme#25]
There is a problem when trying to deduce types 
in a nested IR structure. Imagine you have a derived 
oeprator that looks like this:

```
let bla = struct { threadgroup x } in 
( 'a arg )->unit { 
   arg(); 
   x = parallel(job([1-1]) { arg(); });   
}
```

The type 'a is deduced in a correct way in the outer call of arg().
Unfortunately the type information is not forwarded into the 
parallel or job construct. This will introduce a semantic error
because the type of arg() in the parallel job is 'a and not known.

**Comment by PeterTh** [comment]
Please find out if this issue is specific to JobExprs or more general.

**Comment by HerbertJordan** [comment]
How do you define "correctly deduced"? 

Are you referring to the backend being unable to specialize generic types into concrete C types?

**Comment by stefanmoosbrugger** [comment]
No. This bug does not affect the backend. Just imagine we have a derived operator
that takes an argument of type 'a and we want to create a new variable of type 'a 
inside of our parallel job and another one that is declared outside of the 
parallel job. This thing would somehow look like:
```
LANG_EXT_DERIVED(WhateverIR, 
"('a arg)->unit {
    decl 'a v1...;
    parallel(job([1-1]) { decl 'a v2 ... });
}");
``` 
After our call to someIRExtension.getWhateverIR(refInt4Variable) we 
get an IR construct that would look like:
```
(ref<int<4>> arg)->unit {
    decl ref<int<4>> v1...;
    parallel(job([1-1]) { decl ref<'a> v2 ... });
}
``` 
As you can see, the type of the argument and the outer declaration is 
not 'a anymore. It is ref<int<4>> which is fine. The variable declaration 
inside the parallel job is still 'a which is not correct because it should 
also be ref<int<4>>.

**Comment by HerbertJordan** [comment]
But the `someIRExtension.getWhateverIR` implementation is a manual overload. Usually the getter functions do not accept any arguments. The result of those should contain 'a variables outside and within the job expression. What utility is utilized to get the specialized version where the outer 'a is replaced by `ref<int<4>>`?

**Comment by stefanmoosbrugger** [comment]
Oh sorry, you're completely right. Just tried it out and remembered that the IR construct is built like ```builder.callExpr(someIRExtension.getWhateverIR(), refIntVar)```. The result is a call expression to a 'a "version" of the construct. I guess this is correct, right?! 

The wrong type deduction occurs when utilizing:
```
// extract target function
core::ExpressionPtr fun = core::analysis::stripAttributes(call->getFunctionExpr());
// compute substitutions
core::types::SubstitutionOpt&& map = core::types::getTypeVariableInstantiation(manager, call);
// instantiate type variables according to map
auto lambda = core::transform::instantiate(manager, fun.as<core::LambdaExprPtr>(), map); 
```

The type of the outer 'a var of ```lambda``` is substituted correctly but the 'a var that is contained in the job expression is still of type 'a. I don't know if this bug belongs to the backend or the core (at least the lambda instatiation is done in the backend). If yes, feel free to re-add the backend tag.

**Comment by HerbertJordan** [comment]
Can you do me a favor and post a pretty print of a minimal example of such a call? 

The job expressions as used by the parser is syntactic sugar and I'd like to see what it is actually turned into but I lack the resources to run it myself on my laptop.

**Comment by stefanmoosbrugger** [comment]
Ok, no problem.

Minimal derived operator:
```
LANG_EXT_DERIVED(Test,
    "('a f) -> unit {  "
    "    auto y = f;
    "    let x = ('a a)->unit { auto z = a; };  "
    "    x(f);  "
    "}"
);
```

After calling the getter for this derived operator the result looks like this:
-Pretty print:
```
let fun000 = fun('a v1) -> unit {
    decl 'a v2 = v1;
};

Test(v6)
```
-Detail print:
```
let fun000 = $[NormalizeAnnotation, Value(NameTag): fun('a $[Value(NameTag): v1]$) -> unit {
    decl 'a $[Value(NameTag): v2]$ = $[Value(NameTag): v1]$;
}]$;

let fun001 = $[Value(DerivedTag), NormalizeAnnotation: fun('a $[Value(NameTag): v1]$) -> unit {
    decl 'a $[Value(NameTag): v2]$ = $[Value(NameTag): v1]$;
    fun000($[Value(NameTag): v1]$);
}]$;

fun001(v6)
```
Ok, until now everything is fine and looks ok. In the backend this lambda has to be instantiated.
This behaviour can be seen in https://github.com/insieme/insieme/blob/master/code/backend/src/function_manager.cpp#L412-L437

In my test I'm only interested in the instantiated lambda. So lets instantiate the test lambda that we've created before. The pretty printed version looks like this:
```
let fun000 = fun('a v1) -> unit {
    decl 'a v2 = v1;
};

let fun001 = fun(int<4> v1) -> unit {
    decl int<4> v2 = v1;
    fun000(v1);
};

fun001
```
Now the bug can be seen. The declaration of v2 in fun001 is typed to int4. But everything that is inside fun000 is still typed to 'a. 

The reason for this is that the mapper that is replacing the types (e.g., 'a by int4) does not step into lambdas or function types. (see https://github.com/insieme/insieme/blob/master/code/core/src/transform/node_replacer.cpp#L960-L967)

**Comment by HerbertJordan** [comment]
Thx for the prints.

Right, the reason that the 'a are not replaced is, as you said, that the replacer is not stepping inside fun000 -- howerver this is unfortunately also the intended behavior. The 'a in the inner function is an independent variable of the 'a in the outer. The lambda is the binding scope for those. Those are like variables in different scopes.

The thing is, that in this example, once the body of fun001 is processed by the backend and the call to fun000 is reached it will also get instantiated. So this code section should work. 

What I assume your actual problem is based on is that a IR parser-job-expression is internally mapped to a combination of a JobExpr, a Bind and a Lambda. And this Lambda does not have any arguments, thus also no generic parameters, yet utilizes a 'a variable in its body -- since it is captured in the creation of the job. So what you actually end up with is a unbound type variable 'a in the body of the Job expression which will then lead to faulty C++ output code. Do you agree?

**Comment by HerbertJordan** [comment]
I think I overlooked that the lambda in the bind is indeed generic. This could provide a cheap solution. However, can you repeat the example above by using a job expression instead of a nested function call?

**Comment by stefanmoosbrugger** [comment]
I aggree. Yes, this is exactly how the IR job expression is mapped. 
Ok. so now lets modify the derived operator a little bit.

New derived operator:
```
"('a f) -> unit { "
"   auto y = f;"
"   parallel(job { auto z = f; }); "
"}"
```

Print before instatiation:
-Detail
```
let fun000 = $[NormalizeAnnotation: fun('a $[Value(NameTag): v1]$) -> unit {
    decl 'a $[Value(NameTag): v2]$ = $[Value(NameTag): v1]$;
}]$;

let fun001 = $[Value(DerivedTag), NormalizeAnnotation: fun('a $[Value(NameTag): v1]$) -> unit {
    decl 'a $[Value(NameTag): v2]$ = $[Value(NameTag): v1]$;
    parallel:$[VariableInstantionInfo: (job) -> threadgroup]$(job(([1:uint<8>-inf])){
        bind(){fun000($[Value(NameTag): v1]$)}
    });
}]$;

fun001(v5)
```
Print after instantiation:
-Pretty:
```
let fun000 = fun('a v1) -> unit {
    decl 'a v2 = v1;
};

let fun001 = fun(int<4> v1) -> unit {
    decl int<4> v2 = v1;
    parallel(job([1-inf]){
        bind(){fun000(v1)}
    });
};

fun001
```

**Comment by HerbertJordan** [comment]
OK, this looks exactly as expected.

Now, parallel jobs are handled in the runtime backend by the WorkItemizer https://github.com/insieme/insieme/blob/master/code/backend/include/insieme/backend/runtime/runtime_preprocessor.h#L64-L69

This itemizer is a pre-processor, which hence runs before the actual conversion and thus instantiation process. As a result, the job body is effectively separated from the call site before the actual conversion happens. To test this you could run it through the sequential backend -- if it works there, we have nailed down the problem. 

To fix it, however, is another issue. The best approach is probably to move the instantiation of generic functions from an on-the-fly handling to a pre-processing phase that has to occure before the work-itemizer is run. This pass must not instantiate generic functions for which specific handlers have been defined (e.g. generic IR core functionality). This should solve the issue.

**Comment by PeterTh** [comment]
So basically you could use new recursive instantiation feature, but need to add an (optional) list of lambdas which should not be instantiated by it to the function. To that list you'll have to (at the call site in the backend) add all lambdas which have defined handlers.

I guess.

**Comment by HerbertJordan** [comment]
Is it dealing with type variables too?

**Comment by HerbertJordan** [comment]
In any way, this should at most require a few lines of code for a pre-order depth first replacer (e.g. with some lambda mapper) + the check for the stop condition and the utilization of one or the other instantiation tool.

**Comment by stefanmoosbrugger** [comment]
Fixed by e6acba4667e345c4a563b2f0d010f957bfcb098c

Ok, the fix for this thing was not that easy and there are some issues that should be checked. For example the instantiator that Peter T. wrote, does not support recursive lambdas (currently) and the instantiator fails if the passed IR construct contains recurisve types. 

In order to implement the type fixing utility, I have to check if the lambda contains recursive elements. If yes, the instantiation is not done. Should be fixed at some point. 

Another thing is that I had to change the node replacer. The node replacer takes a lambda argument now that can be used to check every node if it should be handled or not. So basically the node is passed to the lambda and the lambda is returning a boolean that indicates if it should be skipped or not (default argument of the node replacer method looks like: ```replaceAll(..., std::function<...> [](const NodePtr& ptr) { return false; } ```). Looks a little bit ugly in my opinion. Anyway, the crucial part is that this might be a performance killer. I have not seen any performance impacts in the integration tests or unit tests but I'm afraid that is has some impact at least for the debug build. According to some tests it should not impact the release build performance.
